### PR TITLE
nixos/make-options-doc: render with markdown-it-py

### PR DESCRIPTION
**Pull request by pennae**
###### Description of changes

replace mistune in the nixos manual build with markdown-it-py. markdown-it-py is the closest thing to the MyST parser we can get (and the official MyST parser even uses it as its base). the plugins support myst roles and admonitions, both of which we had to hack into mistune. we can now also support definition lists, which would've been amost impossible to do with mistune. definition lists in particular are necessary to render module manual chapters, and (iwrc) home-manager would like to have them too.

changes to the rendered manual are minimal and mostly restricted to whitespace being slightly different (because mistune ate spaces it should've left alone).

next steps after this is merged: extract the rendering code into a new utility, add some tests, and use it for the module chapters in the manual as well (removing their pandoc-generated xml from the tree). after that we can probably have a good think about removing docbook entirely.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 1]
```diff
@@ -91,11 +91,11 @@ def link(self, link, text=None, title=None):
     def list(self, text, ordered, level, start=None):
         if ordered:
             raise NotImplementedError("ordered lists not supported yet")
-        return f"<itemizedlist>\n{text}\n</itemizedlist>"
+        return f"<para><itemizedlist>\n{text}\n</itemizedlist></para>"
     def list_item(self, text, level):
-        return f"<listitem><para>{text}</para></listitem>\n"
+        return f"<listitem>{text}</listitem>\n"
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/mergeJSON.py]: How do we ensure the content is a valid child? The DocBook model does not allow text elements directly inside [`listitem`](https://tdg.docbook.org/tdg/5.1/listitem.html) and I assume Markdown parser will only insert paragraphs when there are multiple paragraphs in the list item or there are blank lines between the list items.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 2]
```diff
@@ -104,7 +104,7 @@ def admonition(self, text, kind):
         if kind not in admonitions:
             raise NotImplementedError(f"admonition {kind} not supported yet")
         tag = admonitions[kind]
-        return f"<{tag}><para>{text.rstrip()}</para></{tag}>"
+        return f"<para><{tag}>{text.rstrip()}</{tag}></para>"
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/mergeJSON.py]: Is there a reason we need to wrap these elements in paragraphs? IIRC, they can be used stand-alone in most context `para` can.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 3]
```diff
@@ -208,19 +236,37 @@ def optionIs(option: Dict[str, Any], key: str, typ: str) -> bool:
...
+    def convertCode(name: str, option: Dict[str, Any], key: str):
+        if optionIs(option, key, 'literalMD'):
+            option[key] = convertString(name, f"*{key.capitalize()}:*\n{option[key]['text']}")
+        elif optionIs(option, key, 'literalExpression'):
+            code = option[key]['text']
+            if '\n' in code:
+                docbook = convertString(name, f"```\n{code}\n```")
+                option[key] = f"<para><emphasis>{key.capitalize()}:</emphasis></para> {docbook}"
+            else:
+                option[key] = convertString(name, f"*{key.capitalize()}:* ` {code} `")
```
Comment (ncfavier) [inline nixos/lib/make-options-doc/mergeJSON.py:248]:
```suggestion
                option[key] = convertString(name, f"*{key.capitalize()}:* `{code}`")
```

We might also want to check that `code` doesn't contain backticks

**Review by ncfavier — COMMENTED** [review]
HTML diff looks good, code is longer but simpler.

> use it for the module chapters in the manual as well (removing their pandoc-generated xml from the tree)

Not just module chapters, right? We want to do this for the entire NixOS manual?

File: pkgs/development/python-modules/markdown-it-py/default.nix [hunk 4]
```diff
@@ -43,6 +46,7 @@ buildPythonPackage rec {
   checkInputs = [
     psutil
     py
+  ] ++ lib.optionals (! disableTests) [
```
[the remark this answers is not in the document]
Comment (ncfavier) [reply pkgs/development/python-modules/markdown-it-py/default.nix:49]: > a minimum of tests (like import checks) seems sensible to keep as a security net

Fair enough.

> we'd still have to remove these check inputs

Disabling `doCheck` does that automatically.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 5]
```diff
@@ -208,19 +236,48 @@ def optionIs(option: Dict[str, Any], key: str, typ: str) -> bool:
         if '_type' not in option[key]: return False
         return option[key]['_type'] == typ
 
+    def convertCode(name: str, option: Dict[str, Any], key: str):
+        if optionIs(option, key, 'literalMD'):
+            option[key] = convertString(name, f"*{key.capitalize()}:*\n{option[key]['text']}")
```
Comment (ncfavier) [inline nixos/lib/make-options-doc/mergeJSON.py:241]: We're breaking the implicit convention, currently relied upon by nixos-search, that options.json has the same format as what `lib.optionAttrSetToDocList` outputs (up to changing a list into a dict indexed by `name`), and making it closer to an internal data structure destined for `options-to-docbook.xsl`.

We should either undo that (IMO best for now), or revert https://github.com/NixOS/nixos-search/pull/460 first.

**Review by ncfavier — CHANGES_REQUESTED** [review]

File: nixos/lib/make-options-doc/mergeJSON.py (hunk 5 above)
Comment (ncfavier) [reply nixos/lib/make-options-doc/mergeJSON.py:241]: > does nixos-search inspect relatedPackages as well?

Nope

Comment (ncfavier) [reply nixos/lib/make-options-doc/mergeJSON.py]: > once we get rid of options-to-docbook.xsl

What's the plan for that? Wouldn't we need to forbid DocBook descriptions / literalDocBook first?

**Review by ncfavier — APPROVED** [review]

**Pull request merged**